internal/revision/parser.go

```diff
@@ -0,0 +1,581 @@
+// Package revision extracts git revision from string
+// More informations about revision : https://www.kernel.org/pub/software/scm/git/docs/gitrevisions.html
+package revision
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"time"
+)
+
+// ErrInvalidRevision is emitted if string doesn't match valid revision
+type ErrInvalidRevision struct {
+	s string
+}
+
+func (e *ErrInvalidRevision) Error() string {
+	return "Revision invalid : " + e.s
+}
+
+// Revisioner represents a revision component
```

Comment:
A revision component is a part of revision, ( we can see how it works here : https://github.com/src-d/go-git/pull/139/files/e29e554853a1850f42f649b96797c5394d65f2fa#diff-0e4bd274259ee67fba0f3f2541704aefR86), so it's all structs defined after this interface (Ref, TildePath, and so on)